lib/config/config-loader.js

```diff
@@ -0,0 +1,696 @@
+// Requirements
+//------------------------------------------------------------------------------
+
+const path = require("node:path");
+const fs = require("node:fs/promises");
+const findUp = require("find-up");
+const { pathToFileURL } = require("node:url");
+const debug = require("debug")("eslint:config-loader");
+const { FlatConfigArray } = require("../config/flat-config-array");
+
+//-----------------------------------------------------------------------------
+// Types
+//-----------------------------------------------------------------------------
+
+/**
+ * @typedef {import("../shared/types").FlatConfigObject} FlatConfigObject
+ * @typedef {import("../shared/types").FlatConfigArray} FlatConfigArray
+ * @typedef {Object} ConfigLoaderOptions
+ * @property {string|false|undefined} configFile The path to the config file to use.
+ * @property {string} cwd The current working directory.
+ * @property {boolean} ignoreEnabled Indicates if ignore patterns should be honored.
+ * @property {FlatConfigArray} [baseConfig] The base config to use.
+ * @property {Array<FlatConfigObject>} [defaultConfigs] The default configs to use.
+ * @property {Array<string>} [ignorePatterns] The ignore patterns to use.
+ * @property {FlatConfigObject|Array<FlatConfigObject>} overrideConfig The override config to use.
+ * @property {boolean} allowTS Indicates if TypeScript configuration files are allowed.
+ */
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+const FLAT_CONFIG_FILENAMES = [
+    "eslint.config.js",
+    "eslint.config.mjs",
+    "eslint.config.cjs"
+];
+
+const TS_FLAT_CONFIG_FILENAMES = [
+    "eslint.config.ts",
+    "eslint.config.mts",
+    "eslint.config.cts"
+];
+
+const importedConfigFileModificationTime = new Map();
+
+/**
+ * Asserts that the given file path is valid.
+ * @param {string} filePath The file path to check.
+ * @returns {void}
+ * @throws {Error} If `filePath` is not a non-empty string.
+ */
+function assertValidFilePath(filePath) {
+    if (!filePath || typeof filePath !== "string") {
+        throw new Error("'filePath' must be a non-empty string");
+    }
+}
+
+/**
+ * Asserts that a configuration exists. A configuration exists if any
+ * of the following are true:
+ * - `configFilePath` is defined.
+ * - `useConfigFile` is defined.
+ * - `overrideConfig` is defined.
+ * @param {string|undefined} configFilePath The path to the config file.
+ * @param {ConfigLoaderOptions} loaderOptions The options to use when loading configuration files.
+ * @returns {void}
+ * @throws {Error} If no configuration exists.
+ */
+function assertConfigurationExists(configFilePath, loaderOptions) {
+
+    const {
+        configFile: useConfigFile,
+        overrideConfig
+    } = loaderOptions;
+
+    if (!configFilePath && useConfigFile !== false && (!overrideConfig || overrideConfig[Symbol.for("eslint:from-cli")])) {
```

Comment:
If the last condition in this check is removed, all tests will still pass.

```suggestion
    if (!configFilePath && useConfigFile !== false) {
```

Is there still a reason for attaching a symbol to `overrideConfig` arrays from the CLI?

Reply:
Hmmm, that is weird. There was definitely at least one test that was failing without this.

Basically, the CLI always passes `overrideConfig`, so I can't rely on it being `undefined` for this check.

Reply:
Yeah. I actually tried refactoring so that the CLI would only pass `overrideConfig` if there were actual overrides, but it quickly turned into a giant mess. Perhaps with more refactoring around it, this can be cleaned up in the future.